File: app/src/main/java/com/example/aem/Transactions/TransactionViewModel.java
package com.example.aem.Transactions;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;

import com.example.aem.AppDatabase;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

public class TransactionViewModel extends AndroidViewModel {
    private final TransactionsDao transactionsDao;
    private final ExecutorService executorService;

    public TransactionViewModel(@NonNull @NotNull Application application) {
        super(application);
        AppDatabase db = AppDatabase.Companion.getInstance(application);
        assert db != null;
        this.transactionsDao = db.transactionsDao();
        this.executorService = Executors.newSingleThreadExecutor();
    }

    public void insertTransaction(TransactionEntity t) {
        executorService.execute(()->transactionsDao.insertTransaction(t));
    }

    public void deleteAllTransactionsByItemId(String item_id){
        executorService.execute(()->transactionsDao.deleteTransactionsByItemId(item_id));
    }

    public ArrayList<TransactionEntity> getAllTransactionsByItemId(String item_id){
        Future<List<TransactionEntity>> f = executorService.submit(()->transactionsDao.loadAllTransactionsByItemId(item_id));
        ArrayList<TransactionEntity> b = new ArrayList<>();
        try {
            b.addAll(f.get(200, TimeUnit.MILLISECONDS));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return b;
    }

    public TransactionEntity getbyTransactionId(String trans_id){
        Future<TransactionEntity> f = executorService.submit(()->transactionsDao.getbyTransactionId(trans_id));
        TransactionEntity b = null;
        try {
            b = f.get(200, TimeUnit.MILLISECONDS);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return b;
    }

    public ArrayList<CategoryTotal> getCategoryTotals(String item_id) {
        Future<List<CategoryTotal>> f = executorService.submit(()->transactionsDao.getCategoryTotals(item_id));
        ArrayList<CategoryTotal> b = new ArrayList<>();
        try {
            b.addAll(f.get(200, TimeUnit.MILLISECONDS));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return b;
    }

    public ArrayList<CategoryTotal> getEntireListCategoryTotals() {
        Future<List<CategoryTotal>> f = executorService.submit(transactionsDao::getEntireCategoryTotals);
        ArrayList<CategoryTotal> b = new ArrayList<>();
        try {
            b.addAll(f.get(200, TimeUnit.MILLISECONDS));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return b;
    }

    public ArrayList<TransactionEntity> getTransByCatAndItemId(String item_id, String category) {
        Future<List<TransactionEntity>> f = executorService.submit(()->transactionsDao.getTransactionsBasedOnCategoryAndItemId(item_id,category));
        ArrayList<TransactionEntity> b = new ArrayList<>();
        try {
            b.addAll(f.get(200, TimeUnit.MILLISECONDS));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return b;
    }
}

File: app/src/main/java/com/example/aem/Transactions/TransactionViewModel.kt
package com.example.aem.Transactions

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import com.example.aem.AppDatabase
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class TransactionViewModel(application: Application) : AndroidViewModel(application) {
    private val transactionDAO: TransactionsDao
    private val executorService: ExecutorService

    init {
        val db: AppDatabase? = AppDatabase.getInstance(application)
        transactionDAO = db!!.transactionsDao()
        executorService = Executors.newSingleThreadExecutor()
    }

    fun insertTransaction(transaction: TransactionEntity) {
        executorService.execute {
            transactionDAO.insertTransaction(transaction)
        }
    }

    fun deleteAllTransactionsByItemId(itemId: String) {
        executorService.execute {
            transactionDAO.deleteTransactionsByItemId(itemId)
        }
    }

    fun getAllTransactionsByItemId(itemId: String): ArrayList<TransactionEntity> {
        val list = arrayListOf<TransactionEntity>()
        val callable = Callable {
            transactionDAO.loadAllTransactionsByItemId(itemId)
        }
        val f = executorService.submit(callable)
        list.addAll(f.get(500, TimeUnit.MILLISECONDS) as List<TransactionEntity>)
        return list
    }

    fun getSingleTransactionByTranId(transId: String): TransactionEntity {
        val future = executorService.submit {
            transactionDAO.getbyTransactionId(transId)
        }
        return future.get(200, TimeUnit.MILLISECONDS) as TransactionEntity
    }

    fun getCategoryTotals(itemId: String): List<CategoryTotal> {
        val callable = Callable {
            transactionDAO.getCategoryTotals(itemId)
        }
        val f = executorService.submit(callable)
        return f.get(500, TimeUnit.MILLISECONDS)
    }

    fun getEntireListCategoryTotals(): ArrayList<CategoryTotal> {
        val list = arrayListOf<CategoryTotal>()
        val callable = Callable {
            transactionDAO.getEntireCategoryTotals()
        }
        val f = executorService.submit(callable)
        list.addAll(f.get(500, TimeUnit.MILLISECONDS) as List<CategoryTotal>)
        return list
    }

    fun getTransByCatAndItemId(itemId: String, category: String): ArrayList<TransactionEntity> {
        val list = arrayListOf<TransactionEntity>()
        val callable = Callable {
            transactionDAO.getTransactionsBasedOnCategoryAndItemId(itemId, category)
        }
        val f = executorService.submit(callable)
        list.addAll(f.get(500, TimeUnit.MILLISECONDS) as List<TransactionEntity>)
        return list
    }

    fun getCategoryTotalsInTimeFrame(start: String, end: String): List<CategoryTotal> {
        val list = arrayListOf<CategoryTotal>()
        val callable = Callable {
            transactionDAO.getCategoryTotalsInATimeFrame(start, end)
        }
        val f = executorService.submit(callable)
        list.addAll(f.get(500, TimeUnit.MILLISECONDS) as List<CategoryTotal>)
        return list
    }
}